### ISSUE TITLE
Review/fix examples for accessibility #a11y

### ISSUE THREAD

**Issue by patrickhlauke** [bigskysoftware/htmx#1431]
With apologies for a rather high-level/vague issue (compared to more specific listings / PRs with proposed fixes), but: having skimmed a few of the examples provided on https://htmx.org/examples/ I'd note that the majority of the ones I've checked seem to lack at least a few basic keyboard accessibility requirements.

For demos involving buttons, there is currently no actual visible focus indication for when a button has/hasn't got focus.

Beyond that, a few specific issues:

* the tab example is not keyboard-accessible (those `<a>` elements, since they lack an actual `href`, don't receive keyboard focus)
* the reordering example can only be operated with the mouse

I was hoping to be able to jump in and evaluate the examples further for screen reader support (and in particular checking that focus is handled correctly/isn't lost and reset when fragments are reloaded/replaced), but these initial keyboard access problems are already an early-warning sign that there's likely further / more fundamental problems here that need to be addressed first, before even going down the SR testing/improvements route.

**Comment by patrickhlauke** [comment]
more bits/context here https://mastodon.social/@sil/110340406008743130 including

* progress bar demo: after activating the button, focus is lost (as the button is replaced wholesale with the progress indicator, removing the currently focused element from under keyboard users' feet, which - in combination with some screen readers - results in focus being reset to the start of the page). this would be a great excuse to show how authors should actually handle focus programmatically in these cases...

**Comment by dz4k** [comment]
Possible mitigations:

- Find/write good JS-based headless UI components that integrate well with htmx (emit events)
- Experiment with implementing keyboard interactions with hx-on
- Log console errors for common pitfalls:
  - if focus moves the the body after swapping
  - if an <a> with htmx attributes lacks an href (we could also automatically add one)
  - ...

**Comment by MarcusMorba** [comment]
I totally agree. Especially all examples with "div-buttons" should use "button"-tag instead or add the required role and aria-attributes. Details and examples can be found on the ARIA Authoring Practices Guide (APG): https://www.w3.org/WAI/ARIA/apg/patterns/button/

**Comment by hibachrach** [comment]
I'd also note that the modal dialog examples fail in at least a couple ways:
- <kbd>Esc</kbd> does not hide the modal dialog
- <kbd>Tab</kbd> navigation is not trapped inside of the dialog
- Initial focus is not set on something sensible (e.g. the exit button)

In terms of impact, note that "examples" are often interpreted by developers as "the right way to do X" or "copy this source code", so this could have consequences for websites built by folks new to HTMX.

**Comment by Telroshan** [comment]
PRs are always welcome to improve documentation btw, if you feel like it!

**Comment by thibaudcolas** [comment]
I’ve been asked multiple times to review the accessibility of htmx-powered UIs and have noticed a few recurring issues so took some time to review the examples, that others here might find interesting: [htmx accessibility gaps: data and recommendations](https://wagtail.org/blog/htmx-accessibility-gaps-data-and-recommendations/). Quick recap for people who don’t want to read the whole thing.

- Lots of examples missing keyboard accessibility considerations as already noted in this thread
- Tabs examples in particular, very far from the [tabs ARIA roles keyboard behavior](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/)
- [Progress bar](https://htmx.org/examples/progress-bar/) demo has a `role="status"` on a heading, which isn’t a supported role for that element

Other examples issues I noted but didn’t investigate too far:

- I believe the [keyboard shortcut example](https://htmx.org/examples/keyboard-shortcuts/) might be missing ARIA for keyboard shortcuts)
- Indicator images don’t have an `alt` attribute (noted in passing on #3354)
- Dark mode button text fail: #3353
- [Inline validation](https://htmx.org/examples/inline-validation/) example missing `aria-invalid` and `aria-describedby` on the validation error message

**Comment by 1cg** [comment]
no longer relevant
